lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -15,26 +15,212 @@
+                        ->getOpResult(0);
+  }
+  return rewriter.create<comb::ConcatOp>(loc, extensionBits, value)
+      ->getOpResult(0);
+}
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ConstantOpLowering : public OpConversionPattern<ConstantOp> {
+  using OpConversionPattern<ConstantOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ConstantOp constOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(constOp,
+                                                constOp.getConstantValue());
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+struct DivOpLowering : public OpConversionPattern<DivOp> {
+  using OpConversionPattern<DivOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(DivOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto lhsType = op.inputs()[0].getType().template cast<IntegerType>();
+    auto rhsType = op.inputs()[1].getType().template cast<IntegerType>();
+    auto targetType = op.result().getType().template cast<IntegerType>();
+
+    // comb.div* needs identical bitwidths for its operands and their result.
+    // Hence, we need to calculate the minimal bitwidth that can be used to
+    // represent the result as well as the operands without precision or sign
+    // loss. The target size only depends on LHS and already handles the edge
+    // cases where the bitwidth needs to be increased by 1. Thus, the targetType
+    // is good enough for both the result as well as LHS.
+    // The bitwidth for RHS is bit tricky. If the RHS is unsigned and we are
+    // about to perform a signed division, then we need one additional bit to
+    // avoid misinterpretation of RHS as a signed value!
+    bool signedDivision = targetType.isSigned();
+    unsigned extendSize = std::max(
+        targetType.getWidth(),
+        rhsType.getWidth() + (signedDivision && !rhsType.isSigned() ? 1 : 0));
+
+    // Extend the operands
+    Value lhsValue = extendTypeWidth(rewriter, loc, adaptor.inputs()[0],
+                                     extendSize, lhsType.isSigned());
+    Value rhsValue = extendTypeWidth(rewriter, loc, adaptor.inputs()[1],
+                                     extendSize, rhsType.isSigned());
+
+    Value divResult;
+    if (signedDivision)
+      divResult = rewriter.create<comb::DivSOp>(loc, lhsValue, rhsValue)
+                      ->getOpResult(0);
+    else
+      divResult = rewriter.create<comb::DivUOp>(loc, lhsValue, rhsValue)
+                      ->getOpResult(0);
+
+    // finally truncate back to the expected result size!
+    Value truncateResult =
+        extractBits(rewriter, loc, divResult, 0, targetType.getWidth());
```

Comment:
Nit: add `/*...=*/` descriptions to arguments (random `0` requires people to jump back to `extractBits` to check what it implies).